datafusion/physical-expr/src/intervals/utils.rs

```diff
@@ -0,0 +1,193 @@
+        convert_interval_bound_to_duration(&interval.upper),
+    ) {
+        Some(Interval::new(lower, upper))
+    } else {
+        None
+    }
+}
+
+/// Converts an [`IntervalBound`] containing a time interval to one containing a `Duration`, if applicable. Otherwise, returns [`None`].
+fn convert_interval_bound_to_duration(
+    interval_bound: &IntervalBound,
+) -> Option<IntervalBound> {
+    match interval_bound.value {
+        ScalarValue::IntervalMonthDayNano(Some(mdn)) => {
+            interval_mdn_to_duration_ns(&mdn).ok().map(|duration| {
+                IntervalBound::new(
+                    ScalarValue::DurationNanosecond(Some(duration)),
+                    interval_bound.open,
+                )
+            })
+        }
+        ScalarValue::IntervalDayTime(Some(dt)) => {
+            interval_dt_to_duration_ms(&dt).ok().map(|duration| {
+                IntervalBound::new(
+                    ScalarValue::DurationMillisecond(Some(duration)),
+                    interval_bound.open,
+                )
+            })
+        }
+        _ => None,
+    }
+}
+
+/// Converts an [`Interval`] of `Duration`s to one of time intervals, if applicable. Otherwise, returns [`None`].
+pub fn convert_duration_type_to_interval(interval: &Interval) -> Option<Interval> {
+    if let (Some(lower), Some(upper)) = (
+        convert_duration_bound_to_interval(&interval.lower),
+        convert_duration_bound_to_interval(&interval.upper),
+    ) {
+        Some(Interval::new(lower, upper))
+    } else {
+        None
+    }
+}
+
+/// Converts an [`IntervalBound`] containing a `Duration` to one containing a time interval, if applicable. Otherwise, returns [`None`].
+fn convert_duration_bound_to_interval(
+    interval_bound: &IntervalBound,
+) -> Option<IntervalBound> {
+    match interval_bound.value {
+        ScalarValue::DurationNanosecond(Some(duration)) => Some(IntervalBound::new(
+            ScalarValue::new_interval_mdn(
+                0,
+                (duration / NANOSECS_IN_ONE_DAY) as i32,
+                duration % NANOSECS_IN_ONE_DAY,
+            ),
+            interval_bound.open,
+        )),
+        ScalarValue::DurationMillisecond(Some(duration)) => Some(IntervalBound::new(
+            ScalarValue::new_interval_dt(
+                (duration / MILLISECS_IN_ONE_DAY) as i32,
+                (duration % MILLISECS_IN_ONE_DAY) as i32,
+            ),
+            interval_bound.open,
+        )),
+        _ => None,
+    }
+}
+
+fn interval_mdn_to_duration_ns(mdn: &i128) -> Result<i64> {
+    let months = mdn >> 96;
+    let days = (mdn & MDN_DAY_MASK) >> 64;
+    let nanoseconds = mdn & MDN_NS_MASK;
+    if months != 0 {
+        return Err(DataFusionError::Internal(
+            "The interval cannot have a non-zero month value for duration convertibility"
+                .to_string(),
+        ));
+    }
+    let duration_ns = days * NANOSECS_IN_ONE_DAY as i128 + nanoseconds;
```

Comment:
FWIW this isn't technically true... DST means some days have only 23 hours...

Reply:
Thanks for reminding me, you're right. We should not shrink the interval when the day field is also non-zero.